### ISSUE TITLE
Different mocks are used for @Mock and @InjectMock when using @TestInstance(TestInstance.Lifecycle.PER_CLASS) of jUnit5

### ISSUE THREAD

**Issue by davinkevin** [mockito/mockito#1437]
Hi, 

I have a problem which seems to be very similar to this one https://github.com/mockito/mockito/issues/1346 but it is happening when I run test with `@TestInstance(TestInstance.Lifecycle.PER_CLASS)` (which is a better thing to do in Kotlin). 

For the following code, the first iteration is working and the following are not because the behavior of another mock is changed 😰(`foo.bar` isn't the same as `bar`) 

```kotlin
class Bar {
    fun call() = 42
}

class Foo(val bar: Bar) {
    fun call() = bar.call() + 23
}

@TestInstance(TestInstance.Lifecycle.PER_CLASS) // If this is remove, all is ok
@ExtendWith(MockitoExtension::class)
class HelloTest {

    @Mock lateinit var bar: Bar
    @InjectMocks lateinit var foo: Foo

    @RepeatedTest(4)
    fun `should return 24`() {
        /* Given */
        whenever(bar.call()).thenReturn(1)
        /* When */
        val r = foo.call()
        /* Then */
        assertThat(r).isEqualTo(24)
    }

    @RepeatedTest(4)
    fun `@InjectMocked foo should have same bar as bar`() {
        assertThat(foo.bar).isSameAs(bar)
    }
}
```

We can also see this in the IDE : 

First time: 
![image](https://user-images.githubusercontent.com/1970922/43034608-e12d695e-8cdf-11e8-8e07-0493e124a023.png)

Second time: 
![image](https://user-images.githubusercontent.com/1970922/43034611-e87d9f62-8cdf-11e8-8ee9-b2ae3ed6288f.png)

Of course, if I remove the `@TestInstance(TestInstance.Lifecycle.PER_CLASS)`, all is now OK.

I'm running on the last version `2.19.1` with jUnit `5.2.0`. You can see the full example here : 
> https://github.com/davinkevin/issue-mockito-junit5-kotlin

**Comment by davinkevin** [comment]
👍 thanks

Le lun. 23 juil. 2018 à 10:26, Serge Bishyr <notifications@github.com> a
écrit :

**Comment by SeriyBg** [comment]
@TimvdLippe @mockitoguy could you please consult me?
The issue here is that we are creating a new mock before each test method even if the test instance is the same but does not update the dependencies inside the `@InjectMocks` object.
One way to fix it is to avoid the new mock instance to be created and just reset the created ones.
Another way is to make sure that the newly created mocks are updated inside the `@InjectMocks` object.

I assume that the first way is preferred one as this looks to be the behavior for TestNG implementation.
Could you please confirm or deny my assumption? Thank you!

**Comment by TimvdLippe** [comment]
> One way to fix it is to avoid the new mock instance to be created and just reset the created ones.

That seems undesired and could lead to other bugs like static state. Let's try to avoid that.

> Another way is to make sure that the newly created mocks are updated inside the @InjectMocks object.

This would severely complicate the logic of `InjectMocks`, which arguably already is way too complicated.

I think the core issue is that we are re-injecting into the same instance. Can we maybe store something like "this instance already had been injected" in the runner and clear that when the test suite has successfully completed?

**Comment by TimvdLippe** [comment]
@SeriyBg I think because of the TestNG implementation this is necessary. Not 100% into TestNG though.

**Comment by SeriyBg** [comment]
@TimvdLippe yes, but for JUnit shouldn't we as well make sure that the mocking in test A does not infect the mocking in test B? If I understand correctly, if the stubbing happens inside the test method, this means that this mock behavior is expected only for this test method.
Am I missing something?

**Comment by TimvdLippe** [comment]
Hm yes that is true. I fear we can not handle this case then, as we do not know if it is a repeated test?

**Comment by SeriyBg** [comment]
@TimvdLippe with repeated test do you mean the test annotated with `@RepeatedTest`? I cannot see how this case makes the difference.
Still, maybe I am missing some important point, but for me looks like the way how it works for TestNG is how it should work for JUnit with `@TestInstance(TestInstance.Lifecycle.PER_CLASS)`

**Comment by SeriyBg** [comment]
Besides that, I guess that it is possible to know in `MockitoExtension` if it the method is annotated with `@RepeatedTest` :)

**Comment by mockitoguy** [comment]
>Besides that, I guess that it is possible to know in MockitoExtension if it the method is annotated with @RepeatedTest :)

Design note + Pull request would be very welcome :)

**Comment by SeriyBg** [comment]
@mockitoguy in the `MockitoExtension` class there is a method `retrieveAnnotationFromTestClasses`  which can be refactored to generic to return Optional of the desired annotation. For now, it's hardcoded to return `@MockitoSettings` annotation.
But I'm missing the point why is `@RepeatedTest` annotation needed, therefore I cannot explain how this annotation can be used from this point.

**Comment by SeriyBg** [comment]
The same way we can find out whether the test class is annotated with `@TestInstance(TestInstance.Lifecycle.PER_CLASS)`

**Comment by sbrannen** [comment]
Since JUnit Jupiter 5.1, you can check the test instance lifecycle via `org.junit.jupiter.api.extension.ExtensionContext.getTestInstanceLifecycle()`.

**Comment by SeriyBg** [comment]
@sbrannen thanx, didn't know that 👍

**Comment by guillaumeyan** [comment]
Another example to demonstrate the problem

https://github.com/guillaumeyan/pr-mockito-perclass

**Comment by Sam-Kruglov** [comment]
Any plans for this? I prefer using `PER_CLASS` but it seems mockito for some reason recreates all `@Mock` fields before each test still while leaving `@InjectMocks` unchanged.
```java
@ExtendWith(MockitoExtension.class)
class MealServiceTest { // PER_CLASS configured via junit properties

    @Mock MealRepo        repo;
    @Mock NutritionClient nutritionClient;

    @InjectMocks MealService service;

    static <T> T withProperty(String property, Object value) {
        return MockitoHamcrest.argThat(Matchers.hasProperty(property, Matchers.is(value)));
    }

    @Nested
    class create {
        String    name;
        Meal.Type type;
        Integer   calories;

        @BeforeEach
        void setUp() {
            name = "burger";
            type = Meal.Type.LUNCH;
            calories = 300;
        }

        void create() {
            service.create(new CreateMealDto(name, type, calories));
        }

        @Test
        void using_given_calories() { // success (given it's ran first)
            //repo=3466, nutritionClient=3552, service=3553
            create();
            verify(repo).save(withProperty("calories", calories));
        }

        @Test
        void given_no_calories__consult_nutrition_api() { // fail: no interactions with repo
            //repo=3796, nutritionClient=3795, service=3553 -- repo and client changed but the service is still the same
            calories = null;
            val suppliedCalories = 250;
            given(nutritionClient.guessCalories(name)).willReturn(suppliedCalories);

            create();

            verify(repo).save(withProperty("calories", suppliedCalories));
        }
    }
}
```

**Comment by slawekjaranowski** [comment]
Last time I did some refresh and refactor in mockito-testng project.

Now mockito-testng simply call `Mockito.mockitoSession()....startMocking();`

There are no any manual management for mocks and other injections.

TestNG has by  design one instance of test class.

In similar case `@InjectMocks` inject correct mock into annotated class.

One case where class with `@InjectMocks` has old value is when we have class with constructor and value is stored in `final` field.

`final` field are not refreshed before next test. 

Another workaround can be set null to field which should have injected mock in some of after method, so class will be created with corrected injections.

**Comment by slawekjaranowski** [comment]
There is similar problem #1417 Allow @InjectMock into static/final fields

**Comment by slawekjaranowski** [comment]
solutions: 
1. make possibility to update final field in @InjectMocks 
2. copy actual value of  @InjectMocks in org.mockito.internal.configuration.InjectingAnnotationEngine#injectCloseableMocks and restore it in returned lambda - looks easy - everyone will have
3. copy actual value of  @InjectMocks in in place where it is really need  - TestNGListener - always, JUnitExteension/Runner after detect if is need

@mockito/developers - what do you think

**Comment by vghero** [comment]
Hey,
today we also ran into this, being confused why it works for the test in isolation but not when running all test together. Then we saw that setting `@TestInstance(TestInstance.Lifecycle.PER_METHOD)` makes it work. Finally ending up here. Is there any progress on this? We also have just ctor injection and in there we set the value into a final field. So code-wise it's a common case. Also if not fixable in the near future, should Mockito better fail on such situations (if detectable) or at least adding a hint to `@InjectMocks` javadoc could help?

**Comment by gbotka** [comment]
I ran into this yesterday, here is my example:

```
@ExtendWith(MockitoExtension.class)
@TestInstance(TestInstance.Lifecycle.PER_CLASS) // If this is remove, all is ok
class BugTest {

    @Mock
    Dependency dependency;

    @InjectMocks
    TestClass testClass;

    public Stream<Arguments> isApplicable() {
        return Stream.of(
            arguments(true),
            arguments(false)
        );
    }

    @ParameterizedTest
    @MethodSource("isApplicable")
    public void isApplicable(final boolean expected) {
        when(dependency.check()).thenReturn(expected);
        assertThat(testClass.check()).isEqualTo(expected);
    }

    public static class Dependency {
        boolean check() {
            return false;
        }
    }

    public static class TestClass {
        private final Dependency dependency;

        public TestClass(Dependency dependency) {
            this.dependency = dependency;
        }

        boolean check() {
            return dependency.check();
        }

    }

}
```
